candid/src/codegen/mod.rs

```diff
@@ -0,0 +1,179 @@
+//! Provide code generation functions and types for generating code from a Candid file.
+//!
+//! Each language will have its own module under this one, and a `idl_to_${language}`
+//! function should be exported for simplicity.
+//!
+//! The LanguageBinding is available to all and can be used to configure generation for
+//! languages that are unsupported.
+use crate::parser::types::{Binding, Dec, FuncType, IDLType, PrimType, TypeField};
+use crate::IDLProg;
+
+pub mod rust;
+pub use rust::idl_to_rust;
+
+type Result<T = ()> = crate::error::Result<T>;
+
+pub trait LanguageBinding {
+    /// Called when the program starts to be analyzed.
+    fn start(&self) -> Result<()> {
+        Ok(())
+    }
+
+    /// Called when the code generation is done, with the output intended.
+    /// The return value of this function will be the actual string outputted.
+    fn done(&self, output: String) -> Result<String> {
+        Ok(output)
+    }
+
+    /// The header string that will be prepended to the output.
+    fn header(&self) -> Result<String> {
+        Ok(String::new())
+    }
+
+    /// The footer string that will be appended to the output.
+    fn footer(&self) -> Result<String> {
+        Ok(String::new())
+    }
+
+    /// This method is called when a type is used as a right hand side. By default
+    /// it forwards it to a specialized method for each type.
+    fn usage(&self, ty: &IDLType) -> Result<String> {
```

Comment:
The `.did` file is all about type declarations. There is no notion of `usage` or `right hand side`. The codegen should have the freedom to define a new type when you feel like to. And the only way you can "use" the definition is by referring to its name. The other "use" cases, e.g. vec, opt, etc, are really inlining the definition as an optimization.

Reply:
See offline discussion.